atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampBackupRunner.java

```diff
@@ -0,0 +1,176 @@
+    }
+
+    /**
+     * Writes a backup of the existing timestamp to the database, if none exists. After this backup, this timestamp
+     * service can no longer be used until a restore. Note that the value returned is the value that was backed up;
+     * multiple calls to this method are safe, and will return the backed up value.
+     *
+     * This method assumes that the timestamp table exists.
+     *
+     * @param defaultValue value to backup if the timestamp table is empty
+     * @return value of the timestamp that was backed up, if applicable
+     */
+    public synchronized long backupExistingTimestamp(long defaultValue) {
+        return clientPool().runWithRetry(client -> {
+            BoundData boundData = getCurrentBoundData(client);
+            byte[] currentBound = boundData.bound();
+            byte[] currentBackupBound = boundData.backupBound();
+
+            if (CassandraTimestampUtils.isValidTimestampData(currentBackupBound)) {
+                // Backup bound has been updated!
+                Preconditions.checkState(!CassandraTimestampUtils.isValidTimestampData(currentBound),
+                        "We had both backup and active bounds readable! This is unexpected; please contact support.");
+                log.info("[BACKUP] Didn't backup, because there is already a backup bound.");
+                return PtBytes.toLong(currentBackupBound);
+            }
+
+            Preconditions.checkState(currentBound == null || CassandraTimestampUtils.isValidTimestampData(currentBound),
+                    "The timestamp is unreadable, though the backup is also unreadable! Please contact support.");
+            byte[] backupValue = MoreObjects.firstNonNull(currentBound, PtBytes.toBytes(defaultValue));
+            Map<String, Pair<byte[], byte[]>> casMap =
+                    ImmutableMap.of(
+                            CassandraTimestampUtils.ROW_AND_COLUMN_NAME,
+                            Pair.create(currentBound, CassandraTimestampUtils.INVALIDATED_VALUE.toByteArray()),
+                            CassandraTimestampUtils.BACKUP_COLUMN_NAME,
+                            Pair.create(currentBackupBound, backupValue));
+
+            ByteBuffer casQueryBuffer = CassandraTimestampUtils.constructCheckAndSetMultipleQuery(casMap);
+
+            CqlResult casResult = executeQueryUnchecked(client, casQueryBuffer);
+            CassandraTimestampUtils.verifyCompatible(casResult, casMap);
+            log.info("[BACKUP] Backed up the value {}", currentBackupBound);
+            return PtBytes.toLong(backupValue);
+        });
+    }
+
+    /**
+     * Restores a backup of an existing timestamp, if possible. Note that this method will throw if the timestamp
+     * backup is unreadable, *and* the current bound is also unreadable.
+     *
+     * This method assumes that the timestamp table exists.
+     */
+    public synchronized void restoreFromBackup() {
```

Comment:
The executeAndVerify parts (after you get the `casMap` in both functions) are the same, so I've extracted that (easy win).

I did a bit more refactoring and unified the logic for the bound readability checks as far as possible, though it might be a little over-engineered. There's a small behaviour change as well in terms of how we handle a restore when there's nothing in the table - previously this would throw, while now I allow this (it's seen as a case where "active is readable, backup is not"). I think it's fine this way as well.